### ISSUE TITLE
x/crypto/ssh: concurrent call ssh.Dial will fail

### ISSUE THREAD

**Issue by wangwd1991** [golang/go#27140]
Please answer these questions before submitting your issue. Thanks!

### What version of Go are you using (`go version`)?
1.8.3

### Does this issue reproduce with the latest release?
https://github.com/golang/crypto.git
commit : 614d502a4dac94afa3a6ce146bd1736da82514c6

### What operating system and processor architecture are you using (`go env`)?
centos 7

### What did you do?

	auth := make([]ssh.AuthMethod, 0)
	auth = append(auth, ssh.Password("123456"))
	config := &ssh.ClientConfig {
		User: "root",
		Auth: auth,
		HostKeyCallback: func(hostname string, remote net.Addr, key ssh.PublicKey) error {
			return nil
		},
		Timeout: 30 * time.Second,
	}
	for i:=0;i<2;i++ {
		go func() {
			client, err := ssh.Dial("tcp", "ip:port", config)
			fmt.Println(err) // the 2th thread (maybe 3th or other litter value) will error
			time.Sleep(5 * time.Second)
			// error is not here, it just a test code
			client.Close()
		}()
	}

### What did you expect to see?
run pass

### What did you see instead?
<nil>
ssh: handshake failed: ssh: unable to authenticate, attempted methods [none password], no supported methods remain

the second thread error , other machines maybe 3 or ...

**Comment by meirf** [comment]
@wangwd1991:

- does [this](https://stackoverflow.com/questions/47102080/ssh-in-go-unable-to-authenticate-attempted-methods-none-no-supported-method) help you?
- can you elaborate what you mean by "the second thread error , other machines maybe 3 or ..." ? please show the exact output.

**Comment by wangwd1991** [comment]
@meirf 
  Thanks for your fast response.
 
1、That doesn't help me.  The single thread, it works fine.

2、Sorry to unclear description.
     “the second thread error , other machines maybe 3 or ..” means the cycle of the 'for' loop.  
     Using different number to reproduce the problem with different addr (ssh server)

      **for i:=0;i<2;i++** { //maybe 3 for dialing other server
	     go func() {


Or. Should I use a '**lock**' to call ssh.Dial then create ssh session to avoid this problem?

**Comment by wangwd1991** [comment]
@crvv 
I run with the result: 
1 success
0 ssh: handshake failed: ssh: unable to authenticate, attempted methods [none password], no supported methods remain

I think you can try with  for i := 0; i < 3; i ++ {  // with 3 or more.

**Comment by wangwd1991** [comment]
And My server /etc/ssh/sshd_config

MaxSession 10
PasswordAuthentication yes

**Comment by hanwen** [comment]
code looks OK to me. You could try to run with the race detector, but can you confirm it fails with all types of SSH servers?

**Comment by wangwd1991** [comment]
@hamaxx 
I increase the number of loop , then  go run -race main.go ? : 

[root@my-server ssh]# go run -race main.go 
15 ssh: handshake failed: read tcp ip1:port1->ip2:port2: read: connection reset by peer
17 ssh: handshake failed: EOF
19 ssh: handshake failed: EOF
9 success
0 success
6 ssh: handshake failed: ssh: unable to authenticate, attempted methods [none password], no supported methods remain
1 ssh: handshake failed: ssh: unable to authenticate, attempted methods [none password], no supported methods remain
4 ssh: handshake failed: ssh: unable to authenticate, attempted methods [none password], no supported methods remain
11 ssh: handshake failed: ssh: unable to authenticate, attempted methods [none password], no supported methods remain
2 ssh: handshake failed: ssh: unable to authenticate, attempted methods [none password], no supported methods remain
8 ssh: handshake failed: ssh: unable to authenticate, attempted methods [none password], no supported methods remain
7 ssh: handshake failed: ssh: unable to authenticate, attempted methods [none password], no supported methods remain
18 ssh: handshake failed: ssh: unable to authenticate, attempted methods [none password], no supported methods remain
16 ssh: handshake failed: ssh: unable to authenticate, attempted methods [password none], no supported methods remain
3 ssh: handshake failed: ssh: unable to authenticate, attempted methods [none password], no supported methods remain
10 ssh: handshake failed: ssh: unable to authenticate, attempted methods [none password], no supported methods remain
5 ssh: handshake failed: ssh: unable to authenticate, attempted methods [none password], no supported methods remain
13 ssh: handshake failed: ssh: unable to authenticate, attempted methods [none password], no supported methods remain
14 ssh: handshake failed: ssh: unable to authenticate, attempted methods [password none], no supported methods remain
12 ssh: handshake failed: ssh: unable to authenticate, attempted methods [none password], no supported methods remain

My env: 
my ssh is OpenSSH_6.6.1p1
Server  :  a VM CentOS Linux release 7.1.1503 (Core).  
               a Physical Machine CentOS Linux release 7.2.1511 (Core). 

What I guess is that,  client send the ssh package to the server quickly, the server will refuse or discard the package?

**Comment by hanwen** [comment]
probably. Maybe there is rate limiting for password logins to avoid brute-force attacks?

**Comment by wangwd1991** [comment]
@hanwen 
Thanks for your reply.

Although, I should use it with thread safe.

Closing this.

**Comment by songtianyi** [comment]
Same problem #32705 when connecting a firewall device.
When connecting the firewall device with a single thread, no matter the frequency you connect, there is only little probability that error happens. BUT when connect concurrently, the error will happen for sure!

**Comment by songtianyi** [comment]
auth fail but returned nothing from server side
See [client_auth.go](https://github.com/golang/crypto/blob/cc06ce4a13d484c0101a9e92913248488a75786d/ssh/client_auth.go#L350)

**Comment by rustyx** [comment]
@songtianyi are you saying that [client_auth.go#L350](https://github.com/golang/crypto/blob/cc06ce4a13d484c0101a9e92913248488a75786d/ssh/client_auth.go#L350) is causing this issue, or just commenting?

**Comment by songtianyi** [comment]
> @songtianyi are you saying that [client_auth.go#L350](https://github.com/golang/crypto/blob/cc06ce4a13d484c0101a9e92913248488a75786d/ssh/client_auth.go#L350) is causing this issue, or just commenting?

Nope. What i'm pointing out here is the code  client_auth.go#L350 does not return any error message.

**Comment by mentaLwz** [comment]
Does this problem have been solved now? Or any other methods to deal with this?
